### ISSUE TITLE
The same code get different exception betwee environmentMode TRACKED_FULL_ASSERT and FULL_ASSERT

### ISSUE THREAD

**Issue by kentbill** [TimefoldAI/timefold-solver#478]
**Describe the bug**
I have a project that uses the Chained Through Time pattern, when I debug it, an exception through from Timefold. But what makes me feel strange is that when I set different environmentMode values(TRACKED_FULL_ASSERT and FULL_ASSERT ), the different exception throws. So I don't know which is the actual exception.

**Expected behavior**
I expect FULL_ ASSERT and TRACKED_FULL_ASSERT will result in the same exception.

**Actual behavior**
environmentMode with FULL_ ASSERT and TRACKED_FULL_ASSERT resulted in different exceptions.

**To Reproduce**
When I set the environmentMode with FULL_ASSERT(SolverConfig.withEnvironmentMode(EnvironmentMode.FULL_ASSERT), or FAST_ASSERT, the exception is below.

```
java.lang.IllegalStateException: UndoMove corruption ([0/0/0/-22560/0/0/0/0/0/0]soft): the beforeMoveScore ([0/0/0/0]hard/[0/0/0/-775620/0/-8/-360/0/0/-3321]soft) is not the undoScore ([0/0/0/0]hard/[0/0/0/-798180/0/-8/-360/0/0/-3321]soft) which is the uncorruptedScore ([0/0/0/0]hard/[0/0/0/-798180/0/-8/-360/0/0/-3321]soft) of the workingSolution.

1) Enable EnvironmentMode TRACKED_FULL_ASSERT
(if you haven't already) to fail-faster in case there's a score corruption or variable listener corruption.
2) Check the Move.createUndoMove(...) method of the moveClass (ChainedSwapMove).
The move (PD04 {PD02} <-> PD02 {RES(D4-002)}) might have a corrupted undoMove (PD02 {RES(D4-002)} <-> PD04 {PD02}).
3) Check your custom VariableListeners (if you have any)
for shadow variables that are used by score constraints that could cause
the scoreDifference ([0/0/0/-22560/0/0/0/0/0/0]soft).

	at ai.timefold.solver.core.impl.score.director.AbstractScoreDirector.assertExpectedUndoMoveScore(AbstractScoreDirector.java:735)
	at ai.timefold.solver.core.impl.localsearch.decider.LocalSearchDecider.doMove(LocalSearchDecider.java:123)
	at ai.timefold.solver.core.impl.localsearch.decider.LocalSearchDecider.decideNextStep(LocalSearchDecider.java:97)
	at ai.timefold.solver.core.impl.localsearch.DefaultLocalSearchPhase.solve(DefaultLocalSearchPhase.java:72)
	at ai.timefold.solver.core.impl.solver.AbstractSolver.runPhases(AbstractSolver.java:82)
	at ai.timefold.solver.core.impl.solver.DefaultSolver.solve(DefaultSolver.java:193)
	at ai.timefold.solver.core.impl.solver.DefaultSolverJob.call(DefaultSolverJob.java:110)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

When I changed the environmentMode to TRACKED_FULL_ASSERT or NON_REPRODUCIBLE, then exception messages were below.

```
java.lang.IllegalStateException: The entity (PD01) has a variable (previousTaskOrResource) with value (PD03) which has a sourceVariableName variable (nextTask) with a value (null) which is not that entity.
Verify the consistency of your input problem for that sourceVariableName variable.
	at ai.timefold.solver.core.impl.domain.variable.inverserelation.SingletonInverseVariableListener.retract(SingletonInverseVariableListener.java:76)
	at ai.timefold.solver.core.impl.domain.variable.inverserelation.SingletonInverseVariableListener.beforeVariableChanged(SingletonInverseVariableListener.java:32)
	at ai.timefold.solver.core.impl.domain.variable.listener.support.VariableChangedNotification.triggerBefore(VariableChangedNotification.java:15)
	at ai.timefold.solver.core.impl.domain.variable.listener.support.VariableChangedNotification.triggerBefore(VariableChangedNotification.java:6)
	at ai.timefold.solver.core.impl.domain.variable.listener.support.AbstractNotifiable.triggerBefore(AbstractNotifiable.java:70)
	at ai.timefold.solver.core.impl.domain.variable.listener.support.VariableListenerNotifiable.notifyBefore(VariableListenerNotifiable.java:26)
	at ai.timefold.solver.core.impl.domain.variable.listener.support.VariableListenerSupport.beforeVariableChanged(VariableListenerSupport.java:161)
	at ai.timefold.solver.core.impl.domain.variable.listener.support.VariableListenerSupport.simulateGenuineVariableChange(VariableListenerSupport.java:262)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at ai.timefold.solver.core.impl.domain.solution.descriptor.SolutionDescriptor.lambda$visitAllEntities$8(SolutionDescriptor.java:910)
	at ai.timefold.solver.core.impl.domain.solution.descriptor.SolutionDescriptor.visitAllEntities(SolutionDescriptor.java:924)
	at ai.timefold.solver.core.impl.domain.solution.descriptor.SolutionDescriptor.visitAllEntities(SolutionDescriptor.java:910)
	at ai.timefold.solver.core.impl.domain.variable.listener.support.VariableListenerSupport.forceTriggerAllVariableListeners(VariableListenerSupport.java:235)
	at ai.timefold.solver.core.impl.domain.variable.listener.support.VariableListenerSupport.recalculateAllShadowVariablesFromScratch(VariableListenerSupport.java:245)
	at ai.timefold.solver.core.impl.score.director.AbstractScoreDirector.assertExpectedUndoMoveScore(AbstractScoreDirector.java:699)
	at ai.timefold.solver.core.impl.localsearch.decider.LocalSearchDecider.doMove(LocalSearchDecider.java:123)
	at ai.timefold.solver.core.impl.localsearch.decider.LocalSearchDecider.decideNextStep(LocalSearchDecider.java:97)
	at ai.timefold.solver.core.impl.localsearch.DefaultLocalSearchPhase.solve(DefaultLocalSearchPhase.java:72)
	at ai.timefold.solver.core.impl.solver.AbstractSolver.runPhases(AbstractSolver.java:82)
	at ai.timefold.solver.core.impl.solver.DefaultSolver.solve(DefaultSolver.java:193)
	at ai.timefold.solver.core.impl.solver.DefaultSolverJob.call(DefaultSolverJob.java:110)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

## Environment

**Timefold Solver Version or Git ref**: 1.5.0

**Output of `java -version`:** 17.0.5

**Output of `uname -a` or `ver`:**

## Additional information

Provide any and all other information which might be relevant to the issue.

**Comment by triceo** [comment]
Thanks for reporting, @kentbill !
The issue will be fixed in 1.6.0, to be released some time in January 2024.

**Comment by kentbill** [comment]
Got it, thank you for your efforts!
